### ISSUE TITLE
[Bug]: Prefill with prompt length == 1 + num_speculative_tokens misclassified as uniform decode → FULL spec-verify cudagraph skips GDN/hybrid recurrent-state write → deterministic garbage (any spec method incl. ngram; v0.25.1 & v0.26.0)

### ISSUE THREAD

**Issue by avesed** [vllm-project/vllm#49918]
### Summary

With speculative decoding enabled, a **prefill whose prompt tokenizes to exactly `1 + num_speculative_tokens` (K+1) tokens** is misclassified as a uniform spec-decode batch by `_is_uniform_decode` (`vllm/v1/worker/gpu_model_runner.py`, shape-only check) and dispatched through the captured **FULL spec-verify CUDA graph**. On models with recurrent state (GDN / Gated DeltaNet hybrids such as Qwen3-Next / Qwen3.5 / Qwen3.6; Mamba-family presumably likewise), that request's recurrent state is **silently never written**, and every subsequent step computes from a zero state → **deterministic garbage output** for that request.

PR #47123 already fixes this (same root cause, opened 2026-06-30, still open — this issue documents the bug it fixes and adds independent root-cause evidence; we are intentionally **not** opening a duplicate PR).

### Root cause (traced on v0.25.1 sources; line numbers close on current main)

1. `GPUModelRunner._is_uniform_decode` classifies by shape only: `max_num_scheduled_tokens == uniform_decode_query_len and num_tokens == num_reqs * (K+1)`. A prefill with a (K+1)-token prompt (or N equal-length such prompts batched) matches exactly. There is no check that the requests are actually decoding.
2. The batch is dispatched with `uniform_decode=True` → the cudagraph dispatcher replays the captured **FULL** spec-verify graph.
3. The GDN metadata builder (`gdn_attn.py`) built **prefill** metadata for this step (`num_prefills > 0`), so the spec persistent-buffer refresh block (gated on spec classification / `num_prefills == 0`) never runs.
4. The captured spec kernels therefore read **capture-time NULL(0) state indices** from the persistent buffers. The NULL guard in `fused_sigmoid_gating_delta_rule_update` (and the conv update equivalent) **skips the recurrent-state write entirely**.
5. The prompt's conv/SSM state is never created; the first verify (and all later steps) read zeros at the correct slots → wrong logits → deterministic garbage.

Instrumented capture-vs-replay probes confirmed each link: the affected prompt's prefill never executes the GDN prefill path (python-side skipped, replayed as the FULL graph); pre-verify state reads are all-zero at the correct slots on the affected server, byte-equal to the prefill's write on an unaffected control server.

### Trigger condition (measured, not theoretical)

Garbage occurs **iff** `prompt_tokens == K+1` for a raw completion (or N equal-length such prompts arriving together). Every one of 6 test prompts garbled at exactly `K = its_token_count − 1` and was clean at every other K ∈ [2,15]:

| prompt (Qwen3.6 tokenizer) | tokens | garbles at K | clean at |
|---|---|---|---|
| 请用中文写一句关于春天的话。 | 8 | 7 | all other K |
| 请用一句话描述大海。 | 5 | 4 | all other K |
| 写一段关于春天的话。 | 6 | 5 (empty output) | all other K |

Scope proven by controls (all official-image or clean-source runs, temp=0, default `FULL_AND_PIECEWISE`):

- **Any spec method**: MTP and `method=ngram` (no draft head loaded at all) produce the same garbage — the defect is prefill-time dispatch, proposer-independent.
- **Any quantization**: reproduced with W4A16 and W8A8 checkpoints (dispatch code is quant-agnostic).
- **Both official images**: `vllm/vllm-openai:v0.25.1` **and** `v0.26.0` reproduce with the same checkpoint and prompt.
- **Batch-independent**: 2 concurrent 8-token prompts at K=7 (16 tokens, 2 reqs) garble identically; concurrency never changes per-request query length.
- **Dense models are immune**: Qwen3-8B (dense) + ngram K=7 with an exactly-8-token prompt is clean — the same mis-dispatch occurs, but spec-verify attention over a fresh sequence is mathematically identical to prefill attention; only recurrent-state models lose state.
- Chat-template traffic is effectively immune (template padding makes `prompt_tokens == K+1` unlikely), which is why this hides well in chat benchmarks and evals.

Note for reproducers: verify the token count against the served model's own tokenizer — the same string tokenizes differently across models (e.g. the trigger above is 8 tokens with the Qwen3.6 tokenizer, 9 with Qwen3-8B's unless the trailing 。 is dropped).

### Minimal reproduction

```bash
vllm serve <any GDN-hybrid model, e.g. Qwen3.6-27B / Qwen3-Next> \
  --speculative-config '{"method":"ngram","num_speculative_tokens":7,"prompt_lookup_max":4,"prompt_lookup_min":2}'
# no draft head needed with ngram

curl -s localhost:8000/v1/completions -H 'Content-Type: application/json' -d '{
  "model": "<model>",
  "prompt": "<any prompt that tokenizes to exactly 8 tokens>",
  "max_tokens": 32, "temperature": 0}'
# → deterministic garbage; the same prompt with one token more/fewer is clean
```

### Workarounds

- `--cudagraph-mode PIECEWISE` (GDN runs eagerly with correct prefill metadata) — verified clean across all repro configs.
- `--enforce-eager`.

### Fix

PR #47123 addresses exactly this (hybrid-gated `force_uniform_decode=False` when the batch contains context requests) and has unit tests. We independently validated an equivalent minimal guard on a v0.25.1 fork — in `_determine_batch_execution_and_padding`, when `force_uniform_decode is None`:

```python
if uniform_decode:
    uniform_decode = bool(
        (input_batch.num_computed_tokens_cpu[:num_reqs]
         >= input_batch.num_prompt_tokens[:num_reqs]).all()
    )
```

(zero-sync numpy views; capture paths pass `force_uniform_decode` and are untouched; also covers the analogous non-spec `q=1` shape collision of a 1-token prompt, cf. #39928/#39945). Post-fix validation: all previously-garbage configs clean; the (K+1)-token prefill dispatches `PIECEWISE, uniform=False`; genuine verify decode still dispatches its FULL uniform graphs; acceptance length (3.78) and decode throughput unchanged.

Happy to share instrumentation logs (dispatch descriptors, capture-vs-replay state checksums) if useful.

### Environment

2× RTX 3090 (SM86), TP=2, CUDA 12.x; reproduced on official `vllm/vllm-openai:v0.25.1` and `v0.26.0` images and on a v0.25.1 source build; GDN-hybrid Qwen3.6-27B (W4A16) and Qwen3.6-35B-A3B (W8A8) targets.

---

AI-assisted: Yes (Claude). Not duplicating an existing issue (none references this bug or PR #47123); intentionally not opening a duplicate PR since #47123 already fixes the same root cause.
